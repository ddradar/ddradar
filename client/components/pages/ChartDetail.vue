<template>
  <section>
    <card :title="chartTitle" :type="cardType">
      <div class="card-content">
        <div class="table-container">
          <b-table
            :data="scores"
            :loading="loading"
            :mobile-cards="false"
            :selected="userScore"
          >
            <template v-slot="props">
              <b-table-column field="name" label="Name">
                <nuxt-link :to="`/users/${props.row.userId}`">
                  {{ props.row.userName }}
                </nuxt-link>
              </b-table-column>
              <b-table-column field="score" label="Score">
                <score-badge
                  :lamp="props.row.clearLamp"
                  :rank="props.row.rank"
                  :score="props.row.score"
                />
              </b-table-column>
              <b-table-column field="exScore" label="EX SCORE">
                {{ props.row.exScore }}
              </b-table-column>
            </template>

            <template v-slot:empty>
              <section class="section">
                <div class="content has-text-grey has-text-centered">
                  <p>Nothing here.</p>
                </div>
              </section>
            </template>
          </b-table>
        </div>
      </div>
      <footer class="card-footer">
        <a class="card-footer-item" @click="launchScoreEditor()">スコア編集</a>
        <a class="card-footer-item" @click="fetchScores(true)">全件表示</a>
      </footer>
    </card>
    <card title="譜面情報" :type="cardType" collapsible>
      <div class="card-content">
        <div class="table-container">
          <table class="table">
            <thead>
              <tr>
                <th>Notes</th>
                <th><abbr title="Freeze Arrow">FA</abbr></th>
                <th><abbr title="Shock Arrow">SA</abbr></th>
              </tr>
            </thead>
            <tbody>
              <tr>
                <td>{{ chart.notes }}</td>
                <td>{{ chart.freezeArrow }}</td>
                <td>{{ chart.shockArrow }}</td>
              </tr>
            </tbody>
          </table>
        </div>
        <div class="table-container">
          <table class="table">
            <thead>
              <tr>
                <th><abbr title="STREAM">STR</abbr></th>
                <th><abbr title="VOLTAGE">VOL</abbr></th>
                <th>AIR</th>
                <th><abbr title="FREEZE">FRE</abbr></th>
                <th><abbr title="CHAOS">CHA</abbr></th>
              </tr>
            </thead>
            <tbody>
              <tr>
                <td>{{ chart.stream }}</td>
                <td>{{ chart.voltage }}</td>
                <td>{{ chart.air }}</td>
                <td>{{ chart.freeze }}</td>
                <td>{{ chart.chaos }}</td>
              </tr>
            </tbody>
          </table>
        </div>
      </div>
    </card>
  </section>
</template>

<script lang="ts">
import { Component, Prop, Vue } from 'nuxt-property-decorator'

import ScoreBadge from '~/components/pages/ScoreBadge.vue'
import ScoreEditor from '~/components/pages/ScoreEditor.vue'
import Card from '~/components/shared/Card.vue'
import { UserScore } from '~/types/api/score'
import {
  getDifficultyName,
  getPlayStyleName,
  SongInfo,
  StepChart,
} from '~/types/api/song'

@Component({ components: { Card, ScoreBadge } })
export default class ChartDetailComponent extends Vue {
  @Prop({ required: true, type: Object })
  readonly song!: Omit<SongInfo, 'charts'>

  @Prop({ required: true, type: Object })
  readonly chart!: StepChart

  loading = true
  scores: UserScore[] = []

  get userScore() {
    return this.scores.find(s => s.userId === this.$accessor.user?.id)
  }

  get chartTitle() {
    const shortPlayStyle = getPlayStyleName(this.chart.playStyle)[0] + 'P' // 'SP' or 'DP'
    return `${shortPlayStyle}-${getDifficultyName(this.chart.difficulty)}`
  }

  get cardType() {
    return `is-${getDifficultyName(this.chart.difficulty).toLowerCase()}`
  }

  async fetch() {
    await this.fetchScores()
  }

  launchScoreEditor() {
    this.$buefy.modal.open({
      parent: this,
      component: ScoreEditor,
      props: {
        songId: this.song.id,
        playStyle: this.chart.playStyle,
        difficulty: this.chart.difficulty,
        songData: this.song,
      },
      hasModalCard: true,
      trapFocus: true,
    })
  }

  /** Call Get Chart Score API */
  async fetchScores(fetchAllData: boolean = false) {
    this.loading = true
    const playStyle = this.chart.playStyle
    const difficulty = this.chart.difficulty
    const query = fetchAllData ? '?scope=full' : ''
    try {
      this.scores = await this.$http.$get<UserScore[]>(
        `/api/v1/scores/${this.song.id}/${playStyle}/${difficulty}${query}`
      )
    } catch {}
    this.loading = false
  }
}
</script>
