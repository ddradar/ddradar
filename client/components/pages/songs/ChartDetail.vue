<template>
  <section>
    <score-board :chart="chart" :info="song" :open="open" />
    <card title="Chart Info" :type="cardType" collapsible>
      <div class="card-content">
        <div class="content">
          <ul>
            <li><em>Notes</em>: {{ chart.notes }}</li>
            <li><em>Freeze Arrow</em>: {{ chart.freezeArrow }}</li>
            <li><em>Shock Arrow</em>: {{ chart.shockArrow }}</li>
          </ul>
        </div>
        <div class="table-container">
          <table class="table">
            <thead>
              <tr>
                <th><abbr title="STREAM">STR</abbr></th>
                <th><abbr title="VOLTAGE">VOL</abbr></th>
                <th>AIR</th>
                <th><abbr title="FREEZE">FRE</abbr></th>
                <th><abbr title="CHAOS">CHA</abbr></th>
              </tr>
            </thead>
            <tbody>
              <tr>
                <td>{{ chart.stream }}</td>
                <td>{{ chart.voltage }}</td>
                <td>{{ chart.air }}</td>
                <td>{{ chart.freeze }}</td>
                <td>{{ chart.chaos }}</td>
              </tr>
            </tbody>
          </table>
        </div>
      </div>
    </card>
  </section>
</template>

<script lang="ts">
import type { Api, Database } from '@ddradar/core'
import { Song } from '@ddradar/core'
import { Component, Prop, Vue } from 'nuxt-property-decorator'

import Card from '~/components/shared/Card.vue'
import ScoreBoard from '~/components/shared/ScoreBoard.vue'

@Component({ components: { Card, ScoreBoard } })
export default class ChartDetailComponent extends Vue {
  @Prop({ required: true, type: Object })
  readonly song!: Omit<Api.SongInfo, 'charts'>

  @Prop({ required: true, type: Object })
  readonly chart!: Database.StepChartSchema

  @Prop({ required: false, type: Boolean, default: false })
  readonly open!: boolean

  get cardType() {
    return `is-${
      Song.difficultyMap
        .get(this.chart.difficulty)!
        .toLowerCase() as Lowercase<Song.DifficultyName>
    }` as const
  }
}
</script>
