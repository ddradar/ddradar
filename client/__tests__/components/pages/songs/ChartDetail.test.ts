import { testSongData } from '@core/__tests__/data'
import { createLocalVue, mount, RouterLinkStub } from '@vue/test-utils'
import Buefy from 'buefy'

import ChartDetail from '~/components/pages/songs/ChartDetail.vue'

const localVue = createLocalVue()
localVue.use(Buefy)

describe('/components/pages/songs/ChartDetail.vue', () => {
  const song = { ...testSongData, charts: undefined }
  const chart = { ...testSongData.charts[0] }
  const propsData = { song, chart }
  const stubs = { NuxtLink: RouterLinkStub, ScoreBoard: true }

  describe('snapshot test', () => {
    test('renders correctly', () => {
      // Arrange - Act
      const wrapper = mount(ChartDetail, { localVue, propsData, stubs })

      // Assert
      expect(wrapper).toMatchSnapshot()
    })
  })
})
