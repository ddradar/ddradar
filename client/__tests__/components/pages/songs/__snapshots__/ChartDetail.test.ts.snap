// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/components/pages/songs/ChartDetail.vue snapshot test renders correctly 1`] = `
<section>
  <scoreboard-stub info="[object Object]" chart="[object Object]"></scoreboard-stub>
  <div class="collapse card">
    <div class="collapse-trigger">
      <header role="button" aria-controls="Chart Info" class="card-header is-beginner">
        <div class="card-header-title">Chart Info</div> <a class="card-header-icon"><span class="icon"><i class="mdi mdi-menu-up mdi-24px"></i></span></a>
      </header>
    </div>
    <transition-stub name="slide">
      <div id="Chart Info" class="collapse-content" style="display: none;">
        <div class="card-content">
          <div class="content">
            <ul>
              <li><em>Notes</em>: 138</li>
              <li><em>Freeze Arrow</em>: 0</li>
              <li><em>Shock Arrow</em>: 0</li>
            </ul>
          </div>
          <div class="table-container">
            <table class="table">
              <thead>
                <tr>
                  <th><abbr title="STREAM">STR</abbr></th>
                  <th><abbr title="VOLTAGE">VOL</abbr></th>
                  <th>AIR</th>
                  <th><abbr title="FREEZE">FRE</abbr></th>
                  <th><abbr title="CHAOS">CHA</abbr></th>
                </tr>
              </thead>
              <tbody>
                <tr>
                  <td>29</td>
                  <td>22</td>
                  <td>5</td>
                  <td>0</td>
                  <td>0</td>
                </tr>
              </tbody>
            </table>
          </div>
        </div>
      </div>
    </transition-stub>
  </div>
</section>
`;
